### ISSUE TITLE
Segmentation fault when trying to generate stack trace from withing std::terminate handler

### ISSUE THREAD

**Issue by till213** [jeremy-rifkin/cpptrace#27]
Hello Jeremy,

First let me thank you for your fantastic work and for making it available to us!

I am trying to hunt down a (sparse?) "My app won't launch on Windows 11" issue in my own app, and I suspect a (totally unexpected ;)) exception being thrown during application setup, quite possibly from one of my "noexcept" methods (my app doesn't throw exceptions itself, nor should the underlying Qt framework).

So I stumbled over your Cpptrace library, in order to get some idea from where the (presumed) exception may come from. Most of my own methods are marked "noexcept", which in this case would be convenient, because in this case the std::terminate handler would be called. The compilers that I use (Apple Clang, MinGW GCC on Windows...) do not seem to unwind the stack, but rather "jump right to the std::terminate handler". This would be great, as it would allow me to get the stack trace of the point where the (suspected) exception would be thrown (I am trying to generate a "crash handler" really).

However it seems that cpptrace::print_trace (I am actually interested in cpptrace::generate_trace()) **generates a segmentation fault** when run **inside my own "termination handler"** (as set via std::set_terminate( at the very beginning of my app).

And here is the nasty bit: but (seemingly) only when not run inside a debugger - doh!

Before I go into more details, here is a minimal C++ (using CMake) example that reproduces this issue:

  https://github.com/till213/CpptraceTest/tree/main

- When run from within my IDE (Qt Creator in my case) (and compiled in debug mode) I get the following application output:

First trace:
Stack trace (most recent call first):
#0 0x00000001000031fa in  at :0
Second trace (in excepton handler):
Stack trace (most recent call first):
#0 0x0000000100003259 in  at :0
Third trace (in termination handler):
Stack trace (most recent call first):
#0 0x00000001000032cc in  at :0
#1 0x00007fff203e9306 in  at :0
#2 0x00007fff203e92a8 in  at :0
#3 0x00000001000033ca in  at :0
#4 0x0000000100003407 in  at :0
#5 0x0000000100003268 in  at :0

Note that neither source names nore line numers are properly evaluated. But apart from that the application terminates as expected: specifically the third stack trace was generated without crash within my own terminate handler.

- However when running the same executable straight from the command line I get:

% ./CpptraceTest
First trace:
Stack trace (most recent call first):
#0 0x000000010cde71fa in main at main.cpp:31
Second trace (in excepton handler):
Stack trace (most recent call first):
#0 0x000000010cde7259 in main at main.cpp:37
Third trace (in termination handler):
Stack trace (most recent call first):
**zsh: segmentation fault**  ./CpptraceTest

Here is the relevant code (main.cpp):

```c++
#include <exception>
#include <iostream>

#include <cpptrace/cpptrace.hpp>

namespace
{
    void willThrow()
    {
        throw "My deliberate exception";
    }

    void willBreakMyPromiseNotToThrow() noexcept
    {
        throw "Breaking my promise";
    }

    void handleTerminate()
    {
        std::cout << "Third trace (in termination handler):" << std::endl;
        // BUG: This will cause a segmentation fault (on macOS) when run *outside* of a debugger
        //      (regardless of whether compiled as debug or release)
        //
        //      When run *inside* a debugger then we get past this point (except that we won't
        //      get neither source names nor line numbers - but frame addresses are properly
        //      evaluated (as it seems)
        cpptrace::print_trace();
        std::abort();
    }
}

int main(int argc, char *argv[])
{
    // Setup a terminate handler in order to catch exceptions that
    // were unexpectedly - but in this example totally deliberately - thrown
    // from functions declared as "noexcept"
    std::set_terminate(::handleTerminate);

    std::cout << "First trace:" << std::endl;
    // This works
    cpptrace::print_trace();

    try {
        willThrow();
    } catch (...) {
        std::cout << "Second trace (in excepton handler):" << std::endl;
        // This works as well
        cpptrace::print_trace();
    }

    ::willBreakMyPromiseNotToThrow();

    return 0;
}
```

**Comment by jeremy-rifkin** [comment]
Hello! Thank you for the report and the core dump. And the minimal repro! I'll investigate locally.

**Comment by till213** [comment]
I tried to figure out whether the C++ standard specifies that some (standard library) functions may be forbidden to be called from a std::terminate_handler, but to my best knowledge I could not find any words about such restrictions:

https://en.cppreference.com/w/cpp/error/terminate

(except of course if we are dealing with "bad alloc" exceptions that are thrown when memory is exhausted: then any further attempts to allocate "just one more byte... just one more!" are doomed to miserably fail - but let's declare this case as "should never happen" ;))

As a side-note, I am really interested in point 7) of the above cppreference:

7) a [noexcept specification](https://en.cppreference.com/w/cpp/language/noexcept_spec) is violated (it is implementation-defined whether any stack unwinding is done in this case)

As most of my methods are indeed declared "noexcept" (as my own application does not throw exceptions just yet on its own, nor should the underlying Qt framework - as mentioned previously. And I am blatantly ignoring the "out of memory" scenario... ;)) this would help me getting to know where the (presumed - I expect some "Windows 11 specific protection mechanisms, e.g. due to some missing "manifest declarations" etc.) exception is being thrown. Because all the compilers (clang on macOS, MinGW gcc on Windows) that I tried do indeed *not* unwind the stack before the std::terminate is called.

That's why it would be super useful to me if I could generate the stack trace from within my custom std::terminate handler.

Again, I will post further findings about Linux and Windows 10 once I get to it.

**Comment by jeremy-rifkin** [comment]
Oh cool, good find. I was wondering the same thing about limitations within a std::terminate_handler. From https://godbolt.org/z/1ErP4jxzK it looks like might unwinding still happening on gcc, clang, and mingw (__cxa_throw starts the unwind process), however maybe there's something internally that causes the unwind to not look beyond the noexcept calls. Just a guess, I'm not super familiar with the internals here.

Things look ok for clang and gcc on linux (including clang with libc++), and on mingw but only if I turn off `WIN32_EXECUTABLE TRUE`. 



```
First trace:
Stack trace (most recent call first):
#0 0x00007ff700a91588 in main at C:/Users/rifkin/test-cpptrace/src/main.cpp:46
#1 0x00007ff700a912ed in __tmainCRTStartup at C:/M/mingw-w64-crt-git/src/mingw-w64/mingw-w64-crt/crt/crtexe.c:272
#2 0x00007ff700a91405 in mainCRTStartup at C:/M/mingw-w64-crt-git/src/mingw-w64/mingw-w64-crt/crt/crtexe.c:193
#3 0x00007ff8b20226ac in ?? at C:\Windows\System32\KERNEL32.DLL:0
#4 0x00007ff8b3b0aa67 in ?? at C:\Windows\SYSTEM32\ntdll.dll:0
Second trace (in excepton handler):
Stack trace (most recent call first):
#0 0x00007ff700a9160d in main at C:/Users/rifkin/test-cpptrace/src/main.cpp:53
#1 0x00007ff700a912ed in __tmainCRTStartup at C:/M/mingw-w64-crt-git/src/mingw-w64/mingw-w64-crt/crt/crtexe.c:272
#2 0x00007ff700a91405 in mainCRTStartup at C:/M/mingw-w64-crt-git/src/mingw-w64/mingw-w64-crt/crt/crtexe.c:193
#3 0x00007ff8b20226ac in ?? at C:\Windows\System32\KERNEL32.DLL:0
#4 0x00007ff8b3b0aa67 in ?? at C:\Windows\SYSTEM32\ntdll.dll:0
Third trace (in termination handler):
Stack trace (most recent call first):
#0  0x00007ff700a914f0 in handleTerminate at C:/Users/rifkin/test-cpptrace/src/main.cpp:27
#1  0x00007ff8573450e5 in ?? at C:\msys64\ucrt64\bin\libstdc++-6.dll:0
#2  0x00007ff85744a832 in ?? at C:\msys64\ucrt64\bin\libstdc++-6.dll:0
#3  0x00007ff857345638 in ?? at C:\msys64\ucrt64\bin\libstdc++-6.dll:0
#4  0x00007ff882b5264c in ?? at C:\msys64\ucrt64\bin\libgcc_s_seh-1.dll:0
#5  0x00007ff85744b674 in ?? at C:\msys64\ucrt64\bin\libstdc++-6.dll:0
#6  0x00007ff8b3b53dfe in ?? at C:\Windows\SYSTEM32\ntdll.dll:0
#7  0x00007ff8b3ace455 in ?? at C:\Windows\SYSTEM32\ntdll.dll:0
#8  0x00007ff8b3b04454 in ?? at C:\Windows\SYSTEM32\ntdll.dll:0
#9  0x00007ff8b1524c3b in ?? at C:\Windows\System32\KERNELBASE.dll:0
#10 0x00007ff882b52700 in ?? at C:\msys64\ucrt64\bin\libgcc_s_seh-1.dll:0
#11 0x00007ff85744ae28 in ?? at C:\msys64\ucrt64\bin\libstdc++-6.dll:0
#12 0x00007ff700a914b2 in willBreakMyPromiseNotToThrow at C:/Users/rifkin/test-cpptrace/src/main.cpp:15
#13 0x00007ff700a915c9 in main at C:/Users/rifkin/test-cpptrace/src/main.cpp:60
#14 0x00007ff700a912ed in __tmainCRTStartup at C:/M/mingw-w64-crt-git/src/mingw-w64/mingw-w64-crt/crt/crtexe.c:272
#15 0x00007ff700a91405 in mainCRTStartup at C:/M/mingw-w64-crt-git/src/mingw-w64/mingw-w64-crt/crt/crtexe.c:193
#16 0x00007ff8b20226ac in ?? at C:\Windows\System32\KERNEL32.DLL:0
#17 0x00007ff8b3b0aa67 in ?? at C:\Windows\SYSTEM32\ntdll.dll:0
```



The crash you got that ended with `zsh: segmentation fault`, was that on macos with apple clang? I can test that later to try to reproduce if so. If I can reproduce out side qt creator that'd be awesome, not sure if qt is part of the problem or not.

**Comment by till213** [comment]
"The crash you got that ended with zsh: segmentation fault, was that on macos with apple clang?"

Yes, that was on macOS "Big Sur" (with latest Big Sur updates installed), and the default Apple "clang" compiler.

I was just about to come here and confirm that running the same code on Linux Ubuntu 22.04 (with whatever "default C++" compiler it comes, I believe gcc (?)) the code works as expected, that is I am properly getting the third stack trace print, as produced within the "terminate handler".

In fact, I just tried to build a Windows console app (by setting WIN32_EXECUTABLE to FALSE - that was still a "standard setting leftover" from the original "Qt Widget default CMake setup", and the code also works. Again, especially also when run "outside" of some debugger (that is, by simply starting the app from the console, or by double-clickin on it etc.).

I tried both with the MinGW compiler that comes with the Qt framework (open source) installer (GCC, I believe - again, I rarely check the used default compilers theses days, for as long as it compiles and works and for as long as they support C++20 ;) - but of course I can lookup and provide this information, should it become helpful later on) as well with the latest Microsoft Visual Studio compiler - both compilers produce working Windows console apps (again, working in the sense that the stack trace gets properly produced and printed in the std::terminate() case).

Summary up to this point:

- Only the macOS console app (this example app) seems to get stuck ("segmentation fault") when trying to produce the stack trace

However you are right, I came across the original problem within my Qt application. There the problem also occurs on Windows in the "windowed" (where WIN32_EXECUTABLE is set to true). But before I jump to any more conclusions I will add another minimal Qt example application to the existing code (I will try and make the Qt dependency optional, that is if Qt is not installed the Qt example app won't be built). I will let you know once that is done (in fact, the initial code in https://github.com/till213/CpptraceTest was alredy such a "minimal Qt Windows app").

In the meantime I have already slightly updated the code, by setting  WIN32_EXECUTABLE to FALSE and by exiting the std::terminate handler with std::exit(-1) instead of std::abort() (as the later still raises a "Unexpected abort" error, at least on Windows - and we don't want to get confused by that (expected) error message).

Thanks you again for taking your time and providing this fabulous library! It was indeed very easy to use and setup (essentially there was nothing to setup, specifically no other 3rd-party dependencies on none of the systems Linux, macOS or Windows - great job!)

**Comment by till213** [comment]
Okay! So I think I got a step closer to the issue. At least on Windows I believe the problem is related to "paths with spaces".

First, I have extended my demo application by adding a "window app" that simply tries to instantiate a "QMainWindow" that - from within its "noexcept" constructor - throws an exception. The custom std::terminate handler then calls "print stack trace" (now implemented in a dynamic "StrackTrace" library, simply in order to come as close to my own setup in my actual application), also calls cpptrace::generate_trace and then simply shows a dialog message:

![grafik](https://github.com/jeremy-rifkin/cpptrace/assets/1205514/ff1e8cb2-9944-401f-9a96-ef487791be81)

Before we go into details, here is the repo with the updated "v0.1" sources:

https://github.com/till213/CpptraceTest (same URL, so just "for the record")

And a "v0.1" release with attached Windows binaries (including the required Qt dependencies and C++ runtime libs):

https://github.com/till213/CpptraceTest/releases/tag/v0.1

(You probably have to "accept the binary from an unknown author" when trying to launch the two *.exes from the ZIP archive)

So if this dialog message shows up after starting the CpptraceWindowTest.exe then all went fine!

However I initially simply could not reproduce the problem (that's why I also added the dynamic library "StackTrace", because I thought that perhaps calling cpptrace from within such a DLL would make the difference).

So I thought to simply provide you with the binaries (together with all required Qt framework depencencies and GCC C++ runtime libs), copied the directory with the binaries onto my desktop (folder), gave it a last check whether I indeed had added all required dependencies...and what? The "window app" did not work anymore? No message dialog showed up anymore.

The plot is thickening...

So I opened the Command prompt, and ran the console app again after having placed it deliberately into a "path with spaces" (my Windows user name also contains spaces), and indeed:

* When the path contains spaces then the called addr2line.exe does not find some expected file (or so it seems):

````
C:\temp\Path With Spaces\CpptraceTest-v0.1>CpptraceConsoleTest.exe
First trace:
Stack frame addresses:
C:/Qt/Tools/mingw1120_64/bin/addr2line.exe: 'C:\temp\Path': No such file
Cpptrace verification failed at C:/dev/Qt/build-CpptraceTest-Desktop_Qt_6_4_3_MinGW_64_bit-Release/_deps/cpptrace-src/src/symbols/symbols_with_addr2line.cpp:432: std::vector<cpptrace::stacktrace_frame> cpptrace::detail::symbolizer::impl::resolve_frames(const std::vector<void*>&)
    primitive_verify(output.size() == entries_vec.size());
```

There are two noteworthy things:

* The path with spaces: C:\temp\Path With Spaces\...
* Cpptrace seems to find the addr2line.exe in a location that is usally not there on "non-development Windows installations"

In fact, the location C:/Qt/Tools/mingw1120_64/bin/ happens to be part of the Qt framework installation. The Qt variant that I have downloaded comes with MinGW which apparently ships an "addr2line.exe".

And I just figured out why cpptrace was apparently able to find this addr2line.exe, as I was running my tests from within the Git Bash shell that apparently has this executable in its PATH - in fact, when trying to repeat this test from the "normal" Windows command promt then the executables work again, as expected - even WITH "spaces in the path":

```
C:\temp\CpptraceTest-v0.1>CpptraceConsoleTest.exe
First trace:
Stack frame addresses:
7ffc29211405
7ff6891a29b6
7ff6891a1394
7ff6891a14e5
7ffc3ef07613
7ffc3f6226b0
Cpptrace output:
Stack trace (most recent call first):
#0 0x00007ffc292114fd in ?? at cygming-crtbeg:0
#1 0x00007ff6891a29b6 in ?? at fake:0
#2 0x00007ff6891a1394 in ?? at crtexe.c:0
etc.
```




Now whether the issues on macOS are related to this I don't know (there my user does not have spaces). But I suggest we focus on the Windows platform, or more specifically, on the "addr2linesexe" in combination with "paths with spaces".

I will dig into the cpptrace code a bit. Also, I am currently a bit unsure why in my actual application the stack trace generation only seems to fail when done in the std::terminate handler (it *does* work in my other try-catch handlers in my main() function), so I am not quite sure whether this addr2line.exe issue with "spaces in path" is simply another issue that I just stumbled across. But I am currently a bit too tired to figure it out just yet, but I believe we do have some clues to follow...

**Comment by jeremy-rifkin** [comment]
Testing on macos I can reproduce the weirdness of segfaulting but only when not run in a debugger. How odd :) Will investigate more.

**Comment by jeremy-rifkin** [comment]
```cpp
rifkins-Mac-mini:build rifkin$ lldb -c /cores/core.1415 
(lldb) target create --core "/cores/core.1415"
Core file '/cores/core.1415' (x86_64) was loaded.
(lldb) bt
* thread #1
  * frame #0: 0x00007ff8024b0adb libsystem_c.dylib`flockfile + 18
    frame #1: 0x00007ff8024b6fdf libsystem_c.dylib`fseek + 55
    frame #2: 0x00000001091d0d8f libStackTrace.0.1.dylib`unsigned int load_bytes<unsigned int, 0>(obj_file=0x0000000000000000, offset=0) at common.hpp:257:5
    frame #3: 0x00000001091d02dd libStackTrace.0.1.dylib`macho_get_text_vmaddr(path="/usr/lib/libc++abi.dylib") at mach-o.hpp:139:22
    frame #4: 0x00000001091d00be libStackTrace.0.1.dylib`cpptrace::detail::get_module_image_base(entry=0x00007f7bc7a04420) at symbols_with_addr2line.cpp:185:29
    frame #5: 0x00000001091d1bce libStackTrace.0.1.dylib`cpptrace::detail::symbolizer::impl::get_addr2line_targets(this=0x00006000027f4000, dlframes=size=7, trace=size=7) at symbols_with_addr2line.cpp:321:73
    frame #6: 0x00000001091d0786 libStackTrace.0.1.dylib`cpptrace::detail::symbolizer::impl::resolve_frames(this=0x00006000027f4000, frames=size=7) at symbols_with_addr2line.cpp:418:42
    frame #7: 0x00000001091d0598 libStackTrace.0.1.dylib`cpptrace::detail::symbolizer::resolve_frames(this=0x00007ff7b6edb8c0, frames=size=7) at symbols_with_addr2line.cpp:451:27
    frame #8: 0x00000001091cc787 libStackTrace.0.1.dylib`cpptrace::generate_trace(skip=0) at cpptrace.cpp:22:58
    frame #9: 0x00000001091cb275 libStackTrace.0.1.dylib`StackTrace::generate() at StackTrace.cpp:12:24
    frame #10: 0x000000010902708b CpptraceConsoleTest`(anonymous namespace)::handleTerminate() at main.cpp:27:9
    frame #11: 0x00007ff80258f4a7 libc++abi.dylib`std::__terminate(void (*)()) + 8
    frame #12: 0x00007ff80258f458 libc++abi.dylib`std::terminate() + 56
    frame #13: 0x000000010902718b CpptraceConsoleTest`__clang_call_terminate + 11
    frame #14: 0x00000001090271c8 CpptraceConsoleTest`(anonymous namespace)::willBreakMyPromiseNotToThrow() at main.cpp:15:9
    frame #15: 0x0000000109027022 CpptraceConsoleTest`main(argc=1, argv=0x00007ff7b6edbb40) at main.cpp:53:5
    frame #16: 0x000000010cb4352e dyld`start + 462
(lldb) 
```

This is helpful, so there's an edge case in the handling of mach-o files that I haven't handled. Super super helpful, thank you :) Windows would be another issue, I can look at that next.

**Comment by jeremy-rifkin** [comment]
I fixed the segfault in 097725350550ff50af3150152ff18ef90caf72c1 and I was able to reproduce the windows issue, very good find with the paths with spaces. Fixed in d72e30af12da31cf60a2cd0701e75b4fb1e80e86. Please let me know if everything is now good to go for you!

**Comment by jeremy-rifkin** [comment]
I'm glad to hear things are now working for you! Thanks so much for your interest in the library, help with diagnosing the issues here, and for the shout out :) Someone suggested I make a list of projects using cpptrace, if I do so I'll be sure to include your project. Good luck with your endeavors!
